Math: Add Auditory functions library

singalsu opened the pull request:
The PR contains one commit for the library and another commit for cmocka tests.

singalsu commented:
It's still work in progress, the Mel filterbank apply needs a test.

lgirdwood commented:
@ShriramShastry pls can you review. Thanks !

ShriramShastry commented:
> @ShriramShastry pls can you review. Thanks !

Sure, I 'II review the PR.
Thank you!!

singalsu commented:
I just updated the PR with unit tests for Mel filterbank with various configurations:

```
sof/build_ut$ ctest --verbose -R auditory
UpdateCTestConfiguration  from :/home/singalsu/work/current/sof/sof/build_ut/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/singalsu/work/current/sof/sof/build_ut/DartConfiguration.tcl
Test project /home/singalsu/work/current/sof/sof/build_ut
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 52
    Start 52: auditory

52: Test command: /home/singalsu/work/current/sof/sof/build_ut/test/cmocka/src/math/auditory/auditory
52: Test timeout computed to be: 10000000
52: 1..10
52: Max absolute error =  3.00 (max  5.00), error RMS =  0.96 (max  3.00)
52: ok 1 - test_mel_filterbank_16_test1
52: Max absolute error =  3.00 (max  5.00), error RMS =  0.88 (max  3.00)
52: ok 2 - test_mel_filterbank_32_test1
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.47 (max  3.00)
52: ok 3 - test_mel_filterbank_16_test2
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.51 (max  3.00)
52: ok 4 - test_mel_filterbank_32_test2
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.32 (max  3.00)
52: ok 5 - test_mel_filterbank_16_test3
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.32 (max  3.00)
52: ok 6 - test_mel_filterbank_32_test3
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.22 (max  3.00)
52: ok 7 - test_mel_filterbank_16_test4
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.37 (max  3.00)
52: ok 8 - test_mel_filterbank_32_test4
52: Max absolute error =  1.00 (max  1.50), error RMS =  0.22 (max  0.50)
52: ok 9 - test_hz_to_mel
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.30 (max  1.50)
52: ok 10 - test_mel_to_hz
52: # ok - tests
1/1 Test #52: auditory .........................   Passed    0.07 sec

The following tests passed:
	auditory

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.07 sec

```

singalsu commented:
The previous push with FFT patches included passed CI, so I removed them from the latest push. CI will pass after #6061 is merged.

singalsu commented on src/include/sof/math/auditory.h:
```diff
@@ -0,0 +1,114 @@
+struct psy_mel_filterbank {
+	int32_t log_mult; /**< Out, QX.Y scale for log, log10, or dB format */
+	int32_t scale; /**< Out, Scale Q16.16 used for Q1.15 triangles */
+	int32_t scale_inv; /**< Out, Inverse scale Q16.16 applied to triangles data */
```
Fields scale and scale_inv can be converted to local variables. The correction for Mel triangles gain is done with scale_log2 now.

singalsu commented:
New push fixed merge conflict in test/cmocka/src/math/CMakeLists.txt, no other changes.

singalsu commented on test/cmocka/src/math/auditory/auditory.c:
```diff
@@ -0,0 +1,427 @@
+#define MEL_FB32_MAX_ERROR_ABS  5.0
+#define MEL_FB32_MAX_ERROR_RMS  3.0
+
+#define DEBUGFILES
```
This needs to be #undef

singalsu commented on test/cmocka/src/math/auditory/ref_auditory.m:
```diff
@@ -0,0 +1,194 @@
+
+function ref_auditory()
+
+% get_ref_hz_to_mel('ref_hz', 'ref_mel', 'ref_revhz', 'ref_hz_to_mel.h');
```
This line should not be commented to generate the test data

singalsu commented on src/math/auditory/auditory.c:
```diff
@@ -0,0 +1,220 @@
+ /* TODO: Used for now to compensate FFT level vs. reference Matlab code for 512 long
+  * FFT. Check FFT lib scaling and make a closed form formula for all FFT lengths.
+  */
+#define FFT_LEVEL_OFFSET_Q16 Q_CONVERT_FLOAT(18.0, 16)
```
Remove, this is no more used.

singalsu commented on src/include/sof/math/auditory.h:
```diff
@@ -0,0 +1,112 @@
+#include <stdint.h>
+
+#define AUDITORY_EPS_Q31	1 /* Smallest nonzero Q1.31 value */
+#define AUDITORY_LOG_2P31_Q26	Q_CONVERT_FLOAT(21.4875625974, 26) /* log(2^31) */
```
No more used

singalsu commented on src/math/auditory/auditory.c:
```diff
@@ -0,0 +1,220 @@
+		return 0;
+
+	/* Log function input is unsigned UQ32.0, output is UQ5.27 */
+	//log_arg = (((int32_t)hz << 16) / 700 + (1L << 16)) << 10; /* Scale by 2^26 */
```
Some old code version in comment, remove

singalsu commented on src/math/auditory/auditory.c:
```diff
@@ -0,0 +1,220 @@
+#include <errno.h>
+#include <stdint.h>
+
+#define EPS_Q31 1
```
Non used #define

singalsu commented on src/math/auditory/auditory.c:
```diff
@@ -0,0 +1,220 @@
+
+#define MEL_MAX_Q2 Q_CONVERT_FLOAT(4358.351140, 2) /* 32767 Hz */
+
+#define LOG_2P31_Q26 Q_CONVERT_FLOAT(21.4876, 26)
```
This is not used

singalsu reviewed (commented):
Found some non-used definitions from old code version

lgirdwood commented:
@wszypelt @lrudyX this PR shouldn't fail on the DMIC test as it's not executed yet on that path. Can you check and confirm if OK to merge. Thanks !

lgirdwood commented on src/math/auditory/auditory.c:
```diff
@@ -0,0 +1,214 @@
+ * @param hz Frequency as Q16.0 Hz, max 32767 Hz
+ * @return Value as Q14.2 Mel, max 4358.4 Mel
+ */
+int16_t psy_hz_to_mel(int16_t hz)
```
qualify for static inline i.e. will it get called a lot in hot code ?

singalsu replied:
No, it's used only initialization. The Mel filterbank apply is the hot part. Did you find other inline candidates?

lgirdwood replied:
no, but these only came to attention as they were small and simple.

lgirdwood reviewed (commented):
LGTM, but lets confirm if some funcs need to be static inline.

wszypelt commented:
@lgirdwood I ran the CI tests again

lgirdwood commented:
@singalsu CI looks good, but some conflicts.

lgirdwood commented:
SOFCI TEST

lgirdwood commented:
license server issues..... re run the CI

singalsu commented on test/cmocka/src/math/auditory/ref_mel_filterbank_16_test1.h:
```diff
@@ -0,0 +1,557 @@
+ * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ */
+
+/* This file was created with script ref_auditory.m */
```
I'll do the same update as I did for matrix cmocka test vectors export. Would path "sof/test/cmocka/m" be a suitable for common Matlab language helper functions?

singalsu replied:
Done now

singalsu commented:
Reviewers, can you please review this.

ShriramShastry commented on test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m:
```diff
@@ -0,0 +1,95 @@
+		error('Invalid mel_log');
+end
+
+if 0
```
can turn off them using `if isequal (variable_enable, flag)`

singalsu replied:
Done, in beginning now plot_triangles and plot_matlab_triangles to enable the plots.

ShriramShastry commented on test/cmocka/src/math/auditory/mfcc_mel_to_hz.m:
```diff
@@ -0,0 +1,13 @@
+
+hz = 700 * (exp(mel / 1126.9941805389) - 1);
+
+end
```
new line error ?

singalsu replied:
yep it was missing

ShriramShastry commented on src/include/sof/math/auditory.h:
```diff
@@ -0,0 +1,111 @@
+	int mel_bins; /**< In, Number of Mel frequency bins */
+	int data_length; /**< Out, Number of int16_t words in triangles data */
+	enum psy_mel_log_scale mel_log_scale; /**< In, LOG, LOG10 or DB to select Mel format */
+	bool slaney; /**< In, Apply Slaney type normalization for Mel filterbank if true */
```
can rename `slaney`  flag?

singalsu replied:
How, would slaney_normalize be better?

ShriramShastry reviewed (commented):
Code looks sound to me !!

andrula-song commented:
LGTM~

singalsu commented:
> LGTM~

Don't hesitate to use the vote green "V" and red "X" buttons since you took the effort of reading the code!

singalsu commented:
SOFCI TEST
